ADAL/src/ADEnrollmentGateway.m

```diff
@@ -110,71 +125,120 @@ + (NSString *)allEnrollmentIds
         return [[dic objectForKey:USER_ID] isEqualToString:userId];
     }
                                                        error:error];
 }
 
-+ (NSString *)enrollmentIdForUserObjectId:(NSString *)userObjectId tenantId:(NSString *)tenantId error:(NSError**) error
++ (NSString *)enrollmentIdForUserObjectId:(NSString *)userObjectId tenantId:(NSString *)tenantId error:(ADAuthenticationError *__autoreleasing *) error
 {
-    return [ADEnrollmentGateway getEnrollmentIDForIdentifier:^BOOL(NSDictionary * dic) {
+    return [ADEnrollmentGateway getEnrollmentIDForIdentifier:^BOOL(NSDictionary *dic) {
         return [[dic objectForKey:OID] isEqualToString:userObjectId] && [[dic objectForKey:TID] isEqualToString:tenantId];
     }
                                                        error:error];
 }
 
-+ (NSString *)enrollmentIdForUniqueAccountId:(NSString *)uniqueAccountId error:(NSError**) error
++ (NSString *)enrollmentIdForUniqueAccountId:(NSString *)uniqueAccountId error:(ADAuthenticationError *__autoreleasing *) error
 {
-    return [ADEnrollmentGateway getEnrollmentIDForIdentifier:^BOOL(NSDictionary * dic) {
+    return [ADEnrollmentGateway getEnrollmentIDForIdentifier:^BOOL(NSDictionary *dic) {
         return [[dic objectForKey:UNIQUE_ACCOUNT_ID] isEqualToString:uniqueAccountId];
     }
                                                        error:error];
 }
 
-+ (NSString *)intuneMamResource:(NSString *)authority error:(NSError**) error
++ (NSString *)enrollmentIdIfAvailable:(ADAuthenticationError *__autoreleasing *) error
 {
-    NSString* resourceJSON = nil;
-
-    @synchronized(self)
-    {
-        if (s_intuneResourceJSON)
-        {
-            resourceJSON = s_intuneResourceJSON;
-        }
+    // this will just return the first enrollment ID
+    return [ADEnrollmentGateway getEnrollmentIDForIdentifier:^BOOL(NSDictionary * __unused dic) {
+        return true;
     }
+                                                       error:error];
+}
+
++ (NSString *)enrollmentIDForUniqueAccountID:(NSString *) homeUserID userID:(NSString *) userID error:(ADAuthenticationError *__autoreleasing *) error
+{
+    NSString *enrollmentID;
+    enrollmentID = homeUserID ? [ADEnrollmentGateway enrollmentIdForUniqueAccountId:homeUserID error:error] : nil;
+    if (enrollmentID)
+        return enrollmentID;
+    
+    enrollmentID = userID ? [ADEnrollmentGateway enrollmentIdForUserId:userID error:error] : nil;
+    if (enrollmentID)
+        return enrollmentID;
+    
+    enrollmentID = [ADEnrollmentGateway enrollmentIdIfAvailable:error];
+    return enrollmentID;
+}
+
++ (NSString *)intuneMAMResourceJSON:(NSString *)authority error:(ADAuthenticationError *__autoreleasing *)error
+{
+    NSString *mamResource = [ADEnrollmentGateway intuneMamResource:authority error:error];
+    NSString *normalizedAuthority = [[NSURL URLWithString:authority] adHostWithPortIfNecessary];
```

Comment:
nit: this is technically not authority, can we rename this variable to host/normalizedHost for clarity?